File: goopylib/maths/packing/packing.py
"""
Module defining rectangle packing algorithms.

See https://pds25.egloos.com/pds/201504/21/98/RectangleBinPack.pdf

"""

from __future__ import annotations

from typing import Callable, NoReturn, Optional, Tuple, List


# Bin shadows the Python built-in 'bin', but until I can think
# of a better name to call it, I'm going to stick with bin.
# pylint: disable=redefined-builtin
# pylint: disable=unsubscriptable-object


class Item:
    """
    Class representing a rectangular object to be packed.
    """

    __slots__ = ["_width", "_height", "_long_side", "_short_side", "_x", "_y", "_rotated", "_id"]

    items: int = 0

    def __init__(self, width: float, height: float) -> None:
        self._width: float = width
        self._height: float = height

        if width > height:
            self._long_side: float = width
            self._short_side: float = height
        else:
            self._long_side: float = height
            self._short_side: float = width

        self._x: float = -1
        self._y: float = -1
        self._rotated = False

        self._id = Item.items
        Item.items += 1

    def rotate(self) -> None:
        """
        Swaps the Item's width and height, ie rotates by 90 degrees.
        """
        self._width, self._height = self._height, self._width
        self._rotated = not self._rotated

    def is_vertical(self) -> bool:
        """
        Returns:
            whether the item is oriented vertically
        """
        return self._height > self._width

    def is_horizontal(self) -> bool:
        """
        Returns:
            whether the item is oriented horizontally
        """
        return self._width > self._height

    def is_rotated(self) -> bool:
        """
        Returns:
            whether the item has been rotated
        """
        return self._rotated

    def area(self) -> float:
        """
        Returns:
            the area of the item
        """
        return self._width * self._height

    def p1(self) -> Tuple[float, float]:
        """
        Returns:
            the bottom-left coordinate of the item
        """
        return self._x, self._y

    def p2(self) -> Tuple[float, float]:
        """
        Returns:
            the top-right coordinate of the item
        """
        return self._x + self._width, self._y + self._height

    @property
    def x(self) -> float:
        """
        Returns:
            the x-position of the item
        """
        return self._x

    @x.setter
    def x(self, value: float) -> NoReturn:
        self._x = value

    @property
    def y(self) -> float:
        """
        Returns:
            the y-position of the item
        """
        return self._y

    @y.setter
    def y(self, value: float) -> NoReturn:
        self._y = value

    @property
    def width(self) -> float:
        """
        Returns:
            the width of the item
        """
        return self._width

    @width.setter
    def width(self, value: float) -> NoReturn:
        raise TypeError("Cannot change item!")

    @property
    def height(self) -> float:
        """
        Returns:
            the height of the item
        """
        return self._height

    @height.setter
    def height(self, value: float) -> NoReturn:
        raise TypeError("Cannot change item!")

    @property
    def long_side(self) -> float:
        """
        Returns:
            the long-side of the item
        """
        return self._long_side

    @long_side.setter
    def long_side(self, value: float) -> NoReturn:
        raise TypeError("Cannot change item!")

    @property
    def short_side(self) -> float:
        """
        Returns:
            the short-side of the item
        """
        return self._short_side

    @short_side.setter
    def short_side(self, value: float) -> NoReturn:
        raise TypeError("Cannot change item!")


class Bin:
    """
    Class representing the maximum area in which to pack Items
    """

    __slots__ = ["_width", "_height", "_items", "_id"]

    bins: int = 0

    def __init__(self, width: float, height: float) -> None:
        self._width: float = width
        self._height: float = height

        self._items: List[Item] = []

        self._id = Bin.bins
        Bin.bins += 1

    def _add(self, item: Item) -> None:
        """
        Adds an item to this bin.
        """
        self._items.append(item)

    def packing_ratio(self) -> float:
        """
        Returns:
            the % packing ratio of this shelf (between 0-1)
        """
        return sum(map(Item.area, self._items)) / (self._width * self._height)

    def items(self) -> List[Item]:
        """
        Returns:
            the items part of the bin
        """
        return self._items.copy()

    @property
    def width(self) -> float:
        """
        Returns:
            the width of the bin
        """
        return self._width

    @width.setter
    def width(self, value: float) -> NoReturn:
        raise TypeError("Cannot change bin!")

    @property
    def height(self) -> float:
        """
        Returns:
            the height of the bin
        """
        return self._height

    @height.setter
    def height(self, value: float) -> NoReturn:
        raise TypeError("Cannot change bin!")


class PackingAlgorithm:
    """
    A class representing a rectangle packing algorithm.

    Args:
        bin_width: the maximum width of the packed bin
        bin_height: the maximum height of the packed bin
    """

    __slots__ = ["_bin_width", "_bin_height", "_bins"]

    def __init__(self, bin_width: float, bin_height: float):
        """
        A class representing a rectangle packing algorithm.

        Args:
            bin_width: the maximum width of the packed bin
            bin_height: the maximum height of the packed bin
        """
        self._bin_width = bin_width
        self._bin_height = bin_height
        self._bins: List[Bin] = []

    def pack(self, item: Item, **kwargs) -> None:
        """
        Packs a rectangular item using this algorithm.

        Args:
            item: the item to pack
            **kwargs:
        """
        raise NotImplementedError()

    def pack_all(self, items: List[Item], sorting: Optional[SortingFunction], **kwargs) -> None:
        """
        Packs a list of rectangular items into bins.

        Args:
            items: the list of items
            sorting: a sorting function to use, otherwise None.
            **kwargs:
        """
        if sorting:
            items = sorting(items)

        for item in items:
            self.pack(item, **kwargs)

    def bins(self) -> Tuple[Bin]:
        """
        Returns:
            a list of bins with packed items.
        """
        return tuple(self._bins)


# Sorting Algorithms


SortingFunction = Callable[[List[Item]], List[Item]]


def sort_by_width(descending: bool = False) -> SortingFunction:
    """
    Args:
        descending: in descending order?

    Returns:
        a function to sort list of items by their width
    """

    def _sort(items: List[Item]):
        return sorted(items, reverse=descending, key=lambda item: item.width)

    return _sort


def sort_by_height(descending: bool = False) -> SortingFunction:
    """
    Args:
        descending: in descending order?

    Returns:
        a function to sort list of items by their height
    """

    def _sort(items: List[Item]):
        return sorted(items, reverse=descending, key=lambda item: item.height)

    return _sort


def sort_by_perimeter(descending: bool = False) -> SortingFunction:
    """
    Args:
        descending: in descending order?

    Returns:
        a function to sort list of items by their perimeter
    """

    def _sort(items: List[Item]):
        return sorted(items, reverse=descending, key=lambda item: item.width + item.height)

    return _sort


def sort_by_area(descending: bool = False) -> SortingFunction:
    """
    Args:
        descending: in descending order?

    Returns:
        a function to sort list of items by their area
    """

    def _sort(items: List[Item]):
        return sorted(items, reverse=descending, key=lambda item: item.width * item.height)

    return _sort


def sort_by_side_ratio(descending: bool = False) -> SortingFunction:
    """
    Args:
        descending: in descending order?

    Returns:
        a function to sort list of items by their side ratio
    """

    def _sort(items: List[Item]):
        return sorted(items, reverse=descending, key=lambda item: item.width / item.height)

    return _sort


def sort_by_long_side(descending: bool = False) -> SortingFunction:
    """
    Args:
        descending: in descending order?

    Returns:
        a function to sort list of items by their long side
    """

    def _sort(items: List[Item]):
        return sorted(items, reverse=descending, key=lambda item: item.long_side)

    return _sort


def sort_by_short_side(descending: bool = False) -> SortingFunction:
    """
    Args:
        descending: in descending order?

    Returns:
        a function to sort list of items by their short side
    """

    def _sort(items: List[Item]):
        return sorted(items, reverse=descending, key=lambda item: item.short_side)

    return _sort


# pylint: enable=redefined-builtin
# pylint: enable=unsubscriptable-object

File: src/goopylib/maths/packing/Packing.cpp
#define GP_LOGGING_LEVEl 3

#include "Packing.h"

// Item Class
namespace gp::packing {
    uint32_t Item::s_Items = 0;

    Item::Item(float width, float height)
            : m_Width(width),
              m_Height(height),
              m_LongSide(max(width, height)),
              m_ShortSide(min(width, height)),
              m_ID(Item::s_Items) {
        Item::s_Items++;
    }

    void Item::rotate() {
        const float tmp = m_Width;
        m_Width = m_Height;
        m_Height = tmp;

        m_Rotated = !m_Rotated;
    }

    bool Item::isVertical() const {
        return m_Height > m_Width;
    }

    bool Item::isHorizontal() const {
        return m_Width > m_Height;
    }

    bool Item::isRotated() const {
        return m_Rotated;
    }

    float Item::area() const {
        return m_Width * m_Height;
    }

    Point Item::p1() const {
        return {m_X, m_Y};
    }

    Point Item::p2() const {
        return {m_X + m_Width, m_Y + m_Height};
    }

    float Item::getX() const {
        return m_X;
    }

    float Item::getY() const {
        return m_Y;
    }

    float Item::getWidth() const {
        return m_Width;
    }

    float Item::getHeight() const {
        return m_Height;
    }

    float Item::getLongSide() const {
        return m_LongSide;
    }

    float Item::getShortSide() const {
        return m_ShortSide;
    }
}

// Bin Class
namespace gp::packing {
    uint32_t Bin::s_Bins = 0;

    Bin::Bin(float width, float height)
            : m_Width(width),
              m_Height(height),
              m_ID(Bin::s_Bins) {
        Bin::s_Bins++;
    }

    void Bin::add(const shared_ptr<Item> &item) {
        m_Items.push_back(item);
    }

    float Bin::packingRatio() const {
        float sum = 0;
        for (auto &item: m_Items) {
            sum += item->area();
        }

        return sum / (m_Width * m_Height);
    }

    std::vector<shared_ptr<Item>> Bin::items() const {
        std::vector<shared_ptr<Item>> copy = m_Items;
        return copy;
    }

    float Bin::getWidth() const {
        return m_Width;
    }

    float Bin::getHeight() const {
        return m_Height;
    }
}

#define RETURN_SORTING_FUNC return [compare](std::vector<shared_ptr<Item>> items) {std::sort(items.begin(), items.end(), compare); return items;}

// Sorting Algorithms
namespace gp::packing {
    SortingFunction sortByWidth(bool descending) {

        std::function<bool(const shared_ptr<Item>, const shared_ptr<Item>)> compare;

        if (descending)
            compare = [](const shared_ptr<Item> &item1, const shared_ptr<Item> &item2) {
                return item1->getWidth() > item2->getWidth();
            };
        else
            compare = [](const shared_ptr<Item> &item1, const shared_ptr<Item> &item2) {
                return item1->getWidth() < item2->getWidth();
            };

        RETURN_SORTING_FUNC;
    }

    SortingFunction sortByHeight(bool descending) {
        std::function<bool(const shared_ptr<Item>, const shared_ptr<Item>)> compare;

        if (descending)
            compare = [](const shared_ptr<Item> &item1, const shared_ptr<Item> &item2) {
                return item1->getHeight() > item2->getHeight();
            };
        else
            compare = [](const shared_ptr<Item> &item1, const shared_ptr<Item> &item2) {
                return item1->getHeight() < item2->getHeight();
            };

        RETURN_SORTING_FUNC;
    }

    SortingFunction sortByPerimeter(bool descending) {
        std::function<bool(const shared_ptr<Item>, const shared_ptr<Item>)> compare;

        if (descending)
            compare = [](const shared_ptr<Item> &item1, const shared_ptr<Item> &item2) {
                return (item1->getWidth() + item1->getHeight()) >
                       (item2->getWidth() + item2->getHeight());
            };
        else
            compare = [](const shared_ptr<Item> &item1, const shared_ptr<Item> &item2) {
                return (item1->getWidth() + item1->getHeight()) <
                       (item2->getWidth() + item2->getHeight());
            };

        RETURN_SORTING_FUNC;
    }

    SortingFunction sortByArea(bool descending) {
        std::function<bool(const shared_ptr<Item>, const shared_ptr<Item>)> compare;

        if (descending)
            compare = [](const shared_ptr<Item> &item1, const shared_ptr<Item> &item2) {
                return (item1->getWidth() * item1->getHeight()) >
                       (item2->getWidth() * item2->getHeight());
            };
        else
            compare = [](const shared_ptr<Item> &item1, const shared_ptr<Item> &item2) {
                return (item1->getWidth() * item1->getHeight()) <
                       (item2->getWidth() * item2->getHeight());
            };

        RETURN_SORTING_FUNC;
    }

    SortingFunction sortBySideRatio(bool descending) {
        std::function<bool(const shared_ptr<Item>, const shared_ptr<Item>)> compare;

        if (descending)
            compare = [](const shared_ptr<Item> &item1, const shared_ptr<Item> &item2) {
                return (item1->getWidth() / item1->getHeight()) >
                       (item2->getWidth() / item2->getHeight());
            };
        else
            compare = [](const shared_ptr<Item> &item1, const shared_ptr<Item> &item2) {
                return (item1->getWidth() / item1->getHeight()) <
                       (item2->getWidth() / item2->getHeight());
            };

        RETURN_SORTING_FUNC;
    }

    SortingFunction sortByLongSide(bool descending) {
        std::function<bool(const shared_ptr<Item>, const shared_ptr<Item>)> compare;

        if (descending)
            compare = [](const shared_ptr<Item> &item1, const shared_ptr<Item> &item2) {
                return item1->getLongSide() > item2->getLongSide();
            };
        else
            compare = [](const shared_ptr<Item> &item1, const shared_ptr<Item> &item2) {
                return item1->getLongSide() < item2->getLongSide();
            };

        RETURN_SORTING_FUNC;
    }

    SortingFunction sortByShortSide(bool descending) {
        std::function<bool(const shared_ptr<Item>, const shared_ptr<Item>)> compare;

        if (descending)
            compare = [](const shared_ptr<Item> &item1, const shared_ptr<Item> &item2) {
                return item1->getShortSide() > item2->getShortSide();
            };
        else
            compare = [](const shared_ptr<Item> &item1, const shared_ptr<Item> &item2) {
                return item1->getShortSide() < item2->getShortSide();
            };

        RETURN_SORTING_FUNC;
    }
}

// Packing Algorithm Class
namespace gp::packing {
    PackingAlgorithm::PackingAlgorithm(float binWidth, float binHeight)
            : m_BinWidth(binWidth),
              m_BinHeight(binHeight) {
    }
}
